indexers/keyvalue/RedisDBIndexer/tests/test_redisdbindexer.py

```diff
@@ -1,58 +1,205 @@
-from jina.drivers.helper import array2pb
-from jina.proto import jina_pb2, uid
-
 
 from .. import RedisDBIndexer
 
+cur_dir = Path(__file__).parent.absolute()
+
+
+def create_document(doc_id, text, weight, length):
+    d = Document()
+    d._document.id = (str(doc_id) * 16)[:16]
+    d.buffer = text.encode('utf8')
+    d.weight = weight
+    d.length = length
+    return d
+
+
+def get_documents(ids):
+    documents = [
+        [0, 'cat', 0.1, 3],
+        [1, 'dog', 0.2, 3],
+        [2, 'crow', 0.3, 4],
+        [3, 'pikachu', 0.4, 7],
+        [4, 'magikarp', 0.5, 8]
+    ]
+    id_to_document = {d[0]: d for d in documents}
+    data = [
+        (id_to_document[id][0], id_to_document[id], MessageToJson(create_document(*id_to_document[id])).encode('utf-8'))
+        for id in ids
+    ]
+    return zip(*data)
+
+
+def apply_action(idx, action):
+    method, key_to_id = action
+    with idx:
+        keys = key_to_id.keys()
+        ids = key_to_id.values()
+        ids, documents, documents_bytes = get_documents(ids)
+        if method == 'add':
+            idx.add(keys, documents_bytes)
+        elif method == 'delete':
+            idx.delete(keys)
+        elif method == 'update':
+            idx.update(keys, documents_bytes)
+        else:
+            print(f'method {method} is not implemented')
+        idx.touch()
+        idx.save()
+
+
+def apply_actions(save_abspath, index_abspath, actions):
+    for action in actions:
+        idx = BaseIndexer.load(save_abspath)
+        apply_action(idx, action)
+
+    return save_abspath, index_abspath
+
+
+def validate_positive_results(keys, documents, searcher):
+    for key, query_doc in zip(keys, documents):
+        result_doc = Parse(searcher.query(key)[0]['values'].decode('utf8'), Document())
+        assert result_doc.id == str(query_doc[0]) * 16
+        assert result_doc.buffer == query_doc[1].encode('utf8')
+        assert round(result_doc.weight, 5) == query_doc[2]
+        assert result_doc.length == query_doc[3]
+
+
+def validate_negative_results(keys, searcher):
+    for key in keys:
+        result_docs = searcher.query(key)
+        assert result_docs == []
+
+
+def validate_results(save_abspath, results, negative_results):
+    with BaseIndexer.load(save_abspath) as searcher:
+        if results:
+            keys, ids = zip(*[[k, v] for k, v in results.items()])
+            _, documents, _ = get_documents(ids)
+            validate_positive_results(keys, documents, searcher)
+        if negative_results:
+            validate_negative_results(negative_results, searcher)
+
+
+def run_crud_test_for_scenario(indexer, actions, results, no_results, tmpdir):
+    import redis
+    r = redis.Redis(host='0.0.0.0', port=63079, db=0, socket_timeout=10)
+    r.flushdb()
+    indexer.workspace = tmpdir
+    with indexer:
+        indexer.touch()
+        indexer.save()
+    save_abspath = indexer.save_abspath
+    index_abspath = indexer.index_abspath
```

Comment:
refactor this into a separate function and move lines 99-102 directly into `run_crud_test_exception_aware` will make the code easier readable. The call-depths will be reduced by two, while explicit in- and output of the redis setup will be defined.

Reply:
Good idea, I made the changes accordingly.